stochrare/dynamics/diffusion.py

```diff
@@ -167,8 +200,14 @@ def trajectory(self, x0, t0, **kwargs):
         if 'brownian_path' in kwargs:
             tw, w = kwargs.pop('brownian_path')
             dw = np.diff(w, axis=0)
+            deltat = tw[1]-tw[0]
+            ratio = int(np.rint(dt/deltat)) # Both int and rint needed here ?
+            dw = dw[:((num-1)*ratio)] # Trim noise vector if sequence w too long
         else:
+            deltat = kwargs.pop('deltat', dt)
+            ratio = int(np.rint(dt/deltat))
             dw = np.random.normal(0, np.sqrt(dt), size=(num-1, dim))
```

Comment:
I think the variance of the Gaussian increments should be `deltat`, not `dt`.

Reply:
Good catch! Actually, I realise that the `else` branch is not tested. The tests for `DiffusionProcess.trajectory` always set the parameter `brownian_path` when calling `trajectory`.
I don't know how to test the case when the brownian path is generated on the fly though, as in this case the solution is not known (since the brownian path is not).

What if `trajectory` returns `t, x, dw` now ? That would enable testing.
A better approach would be the set the seed to generate the exact random sequence ?

Reply:
I am not sure about returning the Brownian path or the Gaussian increment vector (by the way, notations `w` and `dw` for the two things may be used in an inconsistent manner, I should check). In practice we are never interested in it. In practice, we will also virtually always use the `else` branch. Providing the brownian path is really for the few people who want to make careful tests about numerical schemes and their convergence. For the users I can think of, it will be weird to have to do `t, x, w = trajectory()` (or `t, x, _ =`). It would also break constructs of the form `plt.plot(*trajectory(x0, t0))`.

For testing the branch, like you say, the random seed is automatically set (arbitrarily to 100) if the process has the attribute `__deterministic__ = True`, thanks to the `pseudorand` decorator. Hence, this branch can be tested for a deterministic model because the brownian generated on the fly will always be the same. If we want to access the brownian to allow for comparison with an analytical solution (for the geometric brownian motion, for instance), we can simply call `trajectory` first for a `Wiener` process, then the returned `x` should actually be `w`, the same which will be used to integrate numerically the geometric brownian motion in the test.

Reply:
Yes, I just found out about the `pseudorand` decorator :)

Reply:
Although, when `trajectory` is called without a `brownian_path` argument, it generates the increments directly.
```
dw = np.random.normal(0, np.sqrt(deltat), size=(num-1, dim))
```
I'm not sure how to get the expected solution from the sequence of Gaussian increments?

Reply:
The test for `DiffusionProcess1D.trajectory` does what I was referring to:
```
def test_trajectory(self):
    dt_brownian = 1e-5
    model = diffusion1d.DiffusionProcess1D(lambda x, t: 2*x, lambda x, t: x, deterministic=True)
    wiener = diffusion1d.Wiener1D(D=0.5, deterministic=True)
    brownian_path = wiener.trajectory(0., 0., T=0.1, dt=dt_brownian)
    traj_exact = np.exp(1.5*brownian_path[0]+brownian_path[1])
    traj1 = model.trajectory(1., 0., T=0.1, dt=dt_brownian, brownian_path=brownian_path)
    traj2 = model.trajectory(1., 0., T=0.1, dt=dt_brownian, deltat=dt_brownian)
    np.testing.assert_allclose(traj1[1], traj2[1], rtol=1e-5)
    np.testing.assert_allclose(traj1[1], traj_exact, rtol=1e-2)
```
(I have removed the lines corresponding to the test of the Milstein method).

You get the brownian path by calling `wiener.trajectory`. Then, you can provide it explicitly to the `trajectory` method with the `brownian_path` optional argument (`traj1`) or you can generate it on the fly (`traj2`), and the two results coincide, because the realization of the Brownian path is the same (because the model is deterministic).

The function `np.random.normal` is called once in `wiener.trajectory`, then again in the second call of `model.trajectory` defining `traj2`, but the `pseudorand` decorator sets the seed to the same value before calling the function, so `np.random.normal` returns the same thing in both cases.